jl4-lsp/src/LSP/L4/Rules.hs

```diff
@@ -741,3 +723,56 @@ rangeOfResolveWarning = \case
     srcSpanToLspRange Nothing
   Resolve.Ambiguous name _ ->
     srcRangeToLspRange $ rangeOf name
+
+-- ----------------------------------------------------------------------------
+-- Helpers for implementing syntax highlighting
+-- ----------------------------------------------------------------------------
+
+applyPositionMapping :: [SemanticToken] -> PositionMapping -> [SemanticToken]
+applyPositionMapping semTokens positionMapping =
+  Maybe.mapMaybe
+    ( \t ->
+        case toCurrentPosition positionMapping t.start of
+          Nothing -> Nothing
+          Just newPos -> Just (t & #start .~ newPos)
+    )
+    semTokens
+
+-- | @'semanticTokensUsing' phases@
+--
+-- Helper function for defining multi-phase semantic syntax highlighting.
+--
+-- Each phase can produce '[SemanticToken]'s and 'PositionMapping' if the result is outdated.
+-- Tokens obtained from earlier phases take precedence over tokens from later phases.
+--
+-- TODO: this function is inefficient, as all phases are always executed.
+-- It would be better if we could tell whether the result is actually stale and then
+-- we won't run later phases.
+-- E.g. @NormalizedUri -> Action (Maybe ([SemanticToken], Maybe PositionMapping))@
+semanticTokensUsing ::
+  [NormalizedUri -> Action (Maybe ([SemanticToken], PositionMapping))] ->
+  (NormalizedUri -> Action [SemanticToken])
+semanticTokensUsing phases uri = do
+  semTokens <- traverse ($ uri) phases
```

Comment:
why not apply the function to the uri before you hand it to this function? i.e. 
`[Action (Maybe (..)] -> Action [SemanticToken]`?

Reply:
I just thought it looks nicer when calling :shrug:

Reply:
I came up with an artificial reason! It is easier to misuse an API (`[Action (Maybe (..)] -> Action [SemanticToken]`) as there is no guarantee that the `Action` has been called with the same `NormalizedFilePath` :thinking: 

Ok, joking aside, I don't feel too strongly about it, but I do prefer this a little bit.

Reply:
Yeah I'm sure it's fine :)